File: pi/voice_assistant/intent/intent_handler.py
# Convert voice-recognized speech to intents/commands:
#    Ex. "Turn on the bedroom lights" 
#    Output: Intent(action="set", device="bedroom_lights", command="on")

from dataclasses import dataclass
from typing import Optional, List, Dict, Tuple
import re
from enum import Enum
from pathlib import Path
import json
import socket
import time
import paho.mqtt.client as mqtt

MQTT_BROKER_HOST = "127.0.0.1"
MQTT_BROKER_PORT = 1883
TARGET_PLUGS = ("plug1", "plug2")

class IntentAction(Enum): 
    SET = "set"  
    TOGGLE = "toggle"  
    GET = "get"  
    UNKNOWN = "unknown" 

@dataclass
class Intent: 
    action: IntentAction
    device: Optional[str] = None  
    command: Optional[str] = None  
    raw_speech: Optional[str] = None  
    confidence: float = 1.0  

    def __repr__(self):
        return f"Intent(action={self.action.value}, device={self.device}, command={self.command}, confidence={self.confidence:.2f})"

class IntentParser: 
    def __init__(self, config_path: Optional[str] = None, device_aliases: Optional[Dict[str, List[str]]] = None):
        self.device_aliases = {}
        self.action_patterns = []
        
        if config_path is None: 
            config_path = Path(__file__).parent / "config.json"
        else:
            config_path = Path(config_path)
        
        self._load_config(config_path) 
        
        if device_aliases:
            self.device_aliases.update(device_aliases) 

    def _load_config(self, config_path: Path) -> None: 
        try:
            if not config_path.exists():
                self._set_defaults() 
                return
            
            with open(config_path, 'r') as f:
                config = json.load(f)
            
            devices_config = config.get("devices", {}) 
            for device_key, device_info in devices_config.items():
                primary = device_info.get("primary", device_key) 
                aliases = device_info.get("aliases", []) 
                
                self.device_aliases[device_key.lower()] = primary
                for alias in aliases:
                    self.device_aliases[alias.lower()] = primary
            
            actions_config = config.get("actions", [])
            for action_config in actions_config: 
                pattern = action_config.get("pattern")
                action = action_config.get("action")
                command = action_config.get("command")
                
                if pattern and action: 
                    try:
                        intent_action = IntentAction(action) 
                        self.action_patterns.append((pattern, intent_action, command)) 
                    except ValueError:
                        continue
            
            if not self.action_patterns:
                self._set_defaults()
                
        except Exception:
            self._set_defaults()

    def _set_defaults(self) -> None: 
        default_aliases = {
            "bedroom light": "bedroom_light", "bedroom": "bedroom_light",
            "bedroom lights": "bedroom_light", "bedroom lamp": "bedroom_light",
            "plug one": "plug1", "plug 1": "plug1", "first plug": "plug1",
            "plug two": "plug2", "plug 2": "plug2", "second plug": "plug2",
        }
        self.device_aliases = default_aliases
        
        self.action_patterns = [
            (r"turn\s+(?:on|up)\s+(?:the\s+)?(.+)", IntentAction.SET, "on"),
            (r"turn\s+off\s+(?:the\s+)?(.+)", IntentAction.SET, "off"),
            (r"toggle\s+(?:the\s+)?(.+)", IntentAction.TOGGLE, "toggle"),
            (r"(?:what'?s|get|check)\s+(?:the\s+)?(?:status|state)\s+of\s+(?:the\s+)?(.+)", IntentAction.GET, None),
            (r"is\s+(?:the\s+)?(.+)\s+(?:on|off)\??", IntentAction.GET, None),
        ]

    def parse(self, speech: str) -> Intent: 
        if not speech or not isinstance(speech, str):
            return Intent(action=IntentAction.UNKNOWN, raw_speech=speech)

        speech = speech.strip()
        
        for pattern, action, default_command in self.action_patterns:
            match = re.search(pattern, speech, re.IGNORECASE) 
            if match: 
                device_name = match.group(1).strip() if match.groups() else None
                device_id = self._resolve_device(device_name)
                return Intent(
                    action=action, device=device_id, command=default_command,
                    raw_speech=speech, confidence=0.9 if device_id else 0.6  
                )
        return Intent(action=IntentAction.UNKNOWN, raw_speech=speech, confidence=0.0)

    def _resolve_device(self, device_name: Optional[str]) -> Optional[str]: 
        if not device_name: return None
        device_name_lower = device_name.lower().strip()
        if device_name_lower in self.device_aliases:
            return self.device_aliases[device_name_lower]
        for alias, device_id in self.device_aliases.items():
            if device_name_lower in alias or alias in device_name_lower:
                return device_id
        return None

_parser: Optional[IntentParser] = None

def get_parser() -> IntentParser: 
    global _parser
    if _parser is None: _parser = IntentParser()
    return _parser

def std_set_topic(device_id: str) -> str:
    return f"/home/{device_id}/set"

def _infer_target_plug_from_speech(speech: Optional[str]) -> Optional[str]:
    if not speech: return None
    text = speech.lower()
    if re.search(r"\b(plug\s*1|plug\s*one|first\s*plug|plug1)\b", text): return "plug1"
    if re.search(r"\b(plug\s*2|plug\s*two|second\s*plug|plug2)\b", text): return "plug2"
    if re.search(r"\b(plug\s*3|plug\s*three|third\s*plug|plug3)\b", text): return "plug3"
    return None

def publish_intent_to_mqtt(client: mqtt.Client, intent: Intent) -> None:
    if intent.action == IntentAction.SET and intent.command in ("on", "off"):
        command = intent.command
    elif intent.action == IntentAction.TOGGLE:
        command = "toggle"
    else:
        return

    target_device = intent.device if intent.device in TARGET_PLUGS else _infer_target_plug_from_speech(intent.raw_speech)

    if target_device is None:
        print("[MQTT] Skipped publish: no target plug resolved")
        return

    topic = std_set_topic(target_device)
    client.publish(topic, command, qos=0, retain=False)
    print(f"[MQTT] {topic} <- {command}")

def generate_response_text(intent: Intent, target_device: Optional[str]) -> str:
    # This 0.4s file wakes up the physical amplifier on the ESP32!
    base = "silence.mp3,"
    
    # Error states map to a single file
    if intent.action == IntentAction.UNKNOWN:
        return base + "i-m-sorry-i-didn-t-understand-that-command.mp3"
    
    if not target_device:
        return base + "sorry-this-device-is-not-recognized.mp3"
    
    # Determine the target plug file
    device_file = None
    if target_device == "plug1":
        device_file = "plug-one.mp3"
    elif target_device == "plug2":
        device_file = "plug-two.mp3"
    else:
        device_file = "plug-three.mp3"
    #device_file = "plug-one.mp3" if target_device == "plug1" else "plug-two.mp3"
    
    # Combine the action file with the target plug file
    if intent.action == IntentAction.SET:
        if intent.command == "on":
            return base + f"turning-on.mp3,{device_file}"
        else:
            return base + f"turning-off.mp3,{device_file}"
    elif intent.action == IntentAction.TOGGLE:
        return base + f"toggling.mp3,{device_file}"
    elif intent.action == IntentAction.GET:
        return base + f"checking.mp3,{device_file}"
    
    return base + "i-m-sorry-i-didn-t-understand-that-command.mp3"

def run_intent_server(host: str = "127.0.0.1", port: int = 9090) -> None:
    parser = get_parser()
    print(f"Intent server listening on {host}:{port}...")

    mqtt_client = mqtt.Client(client_id=f"intent-handler-{int(time.time())}")
    try:
        mqtt_client.connect(MQTT_BROKER_HOST, MQTT_BROKER_PORT, keepalive=60)
        mqtt_client.loop_start()
        print(f"MQTT connected to {MQTT_BROKER_HOST}:{MQTT_BROKER_PORT}")
    except ConnectionRefusedError:
        print(f"[WARNING] MQTT Broker offline at {MQTT_BROKER_HOST}:{MQTT_BROKER_PORT}. Running without MQTT.")

    try:
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as server:
            server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            server.bind((host, port))
            server.listen()

            while True:
                conn, addr = server.accept()
                with conn:
                    buf = b""
                    while True:
                        data = conn.recv(4096)
                        if not data: break
                        buf += data
                        
                        while b"\n" in buf:
                            line, buf = buf.split(b"\n", 1)
                            if not line.strip(): continue

                            try:
                                payload = json.loads(line.decode("utf-8"))
                                speech = str(payload.get("text", "")).strip()
                                if not speech:
                                    conn.sendall(b"No speech detected.\n")
                                    continue

                                intent = parser.parse(speech)
                                target_device = intent.device if intent.device in TARGET_PLUGS else _infer_target_plug_from_speech(intent.raw_speech)
                                
                                publish_intent_to_mqtt(mqtt_client, intent)
                                
                                # Send the friendly string back to command_listener.py!
                                reply = generate_response_text(intent, target_device)
                                conn.sendall((reply + "\n").encode("utf-8"))
                                
                            except json.JSONDecodeError:
                                conn.sendall(b"I encountered an error parsing the intent.\n")
    finally:
        mqtt_client.loop_stop()
        mqtt_client.disconnect()

if __name__ == "__main__":
    run_intent_server()

File: pi/voice_assistant/intent/intent_handler.cpp
#include <string>
#include <vector>
#include <map>
#include <optional>
#include <regex>
#include <fstream>
#include <filesystem>
#include <iostream>
#include <memory>
#include <nlohmann/json.hpp>

using json = nlohmann::json;
namespace fs = std::filesystem;

// Enum for intent actions
enum class IntentAction {
    SET,
    TOGGLE,
    GET,
    UNKNOWN
};

std::string intentActionToString(IntentAction action) {
    switch (action) {
        case IntentAction::SET:
            return "set";
        case IntentAction::TOGGLE:
            return "toggle";
        case IntentAction::GET:
            return "get";
        case IntentAction::UNKNOWN:
            return "unknown";
        default:
            return "unknown";
    }
}

IntentAction stringToIntentAction(const std::string& str) {
    if (str == "set") return IntentAction::SET;
    if (str == "toggle") return IntentAction::TOGGLE;
    if (str == "get") return IntentAction::GET;
    return IntentAction::UNKNOWN;
}

// Struct for structured intent/command
struct Intent {
    IntentAction action;
    std::optional<std::string> device;
    std::optional<std::string> command;
    std::optional<std::string> raw_speech;
    float confidence;

    Intent(IntentAction a, 
           std::optional<std::string> d = std::nullopt,
           std::optional<std::string> c = std::nullopt,
           std::optional<std::string> r = std::nullopt,
           float conf = 1.0f)
        : action(a), device(d), command(c), raw_speech(r), confidence(conf) {}

    std::string toString() const {
        std::string result = "Intent(action=" + intentActionToString(action);
        
        if (device) {
            result += ", device=" + *device;
        } else {
            result += ", device=None";
        }
        
        if (command) {
            result += ", command=" + *command;
        } else {
            result += ", command=None";
        }
        
        char conf_str[10];
        snprintf(conf_str, sizeof(conf_str), "%.2f", confidence);
        result += ", confidence=" + std::string(conf_str) + ")";
        
        return result;
    }
};

// Action pattern structure
struct ActionPattern {
    std::string pattern;
    IntentAction action;
    std::optional<std::string> command;
};

// Intent Parser class
class IntentParser {
private:
    std::map<std::string, std::string> device_aliases;
    std::vector<ActionPattern> action_patterns;

    // Transform string to lowercase
    static std::string toLower(const std::string& str) {
        std::string result = str;
        std::transform(result.begin(), result.end(), result.begin(),
                      [](unsigned char c) { return std::tolower(c); });
        return result;
    }

    // Trim whitespace from both ends
    static std::string trim(const std::string& str) {
        auto start = str.begin();
        while (start != str.end() && std::isspace(*start)) {
            start++;
        }

        auto end = str.end();
        do {
            end--;
        } while (std::distance(start, end) > 0 && std::isspace(*end));

        return std::string(start, end + 1);
    }

    void loadConfig(const fs::path& config_path) {
        try {
            if (!fs::exists(config_path)) {
                std::cerr << "Warning: Config file not found at " << config_path.string()
                          << ". Using minimal defaults.\n";
                setDefaults();
                return;
            }

            std::ifstream config_file(config_path);
            json config = json::parse(config_file);

            // Load device aliases
            if (config.contains("devices")) {
                const auto& devices_config = config["devices"];
                for (auto& [device_key, device_info] : devices_config.items()) {
                    std::string primary = device_info.value("primary", device_key);
                    std::vector<std::string> aliases = device_info.value("aliases", std::vector<std::string>());

                    device_aliases[toLower(device_key)] = primary;
                    for (const auto& alias : aliases) {
                        device_aliases[toLower(alias)] = primary;
                    }
                }
            }

            // Load action patterns
            if (config.contains("actions")) {
                const auto& actions_config = config["actions"];
                for (const auto& action_config : actions_config) {
                    if (!action_config.contains("pattern") || !action_config.contains("action")) {
                        continue;
                    }

                    std::string pattern = action_config["pattern"];
                    std::string action_str = action_config["action"];
                    std::optional<std::string> command = std::nullopt;

                    if (action_config.contains("command") && !action_config["command"].is_null()) {
                        command = action_config["command"].get<std::string>();
                    }

                    IntentAction intent_action = stringToIntentAction(action_str);
                    if (intent_action == IntentAction::UNKNOWN && action_str != "unknown") {
                        std::cerr << "Warning: Unknown action '" << action_str << "' in config. Skipping pattern.\n";
                        continue;
                    }

                    action_patterns.push_back({pattern, intent_action, command});
                }
            }

            if (action_patterns.empty()) {
                std::cerr << "Warning: No valid action patterns loaded from config. Using defaults.\n";
                setDefaults();
            }

        } catch (const json::exception& e) {
            std::cerr << "Error parsing config JSON: " << e.what() << ". Using defaults.\n";
            setDefaults();
        } catch (const std::exception& e) {
            std::cerr << "Error loading config: " << e.what() << ". Using defaults.\n";
            setDefaults();
        }
    }

    void setDefaults() {
        device_aliases = {
            {"living room light", "living_room_light"},
            {"living room", "living_room_light"},
            {"living room lights", "living_room_light"},
            {"living room lamp", "living_room_light"},
            {"bedroom light", "bedroom_light"},
            {"bedroom", "bedroom_light"},
            {"bedroom lights", "bedroom_light"},
            {"bedroom lamp", "bedroom_light"},
            {"kitchen light", "kitchen_light"},
            {"kitchen", "kitchen_light"},
            {"kitchen lights", "kitchen_light"},
            {"kitchen lamp", "kitchen_light"},
            {"plug one", "plug1"},
            {"plug 1", "plug1"},
            {"first plug", "plug1"},
            {"plug two", "plug2"},
            {"plug 2", "plug2"},
            {"second plug", "plug2"},
        };

        action_patterns = {
            {"turn\\s+(?:on|up)\\s+(?:the\\s+)?(.+)", IntentAction::SET, "on"},
            {"turn\\s+off\\s+(?:the\\s+)?(.+)", IntentAction::SET, "off"},
            {"switch\\s+on\\s+(?:the\\s+)?(.+)", IntentAction::SET, "on"},
            {"switch\\s+off\\s+(?:the\\s+)?(.+)", IntentAction::SET, "off"},
            {"switch\\s+(?:the\\s+)?(.+)\\s+on", IntentAction::SET, "on"},
            {"switch\\s+(?:the\\s+)?(.+)\\s+off", IntentAction::SET, "off"},
            {"toggle\\s+(?:the\\s+)?(.+)", IntentAction::TOGGLE, "toggle"},
            {"flip\\s+(?:the\\s+)?(.+)", IntentAction::TOGGLE, "toggle"},
            {"(?:what'?s|get|check)\\s+(?:the\\s+)?(?:status|state)\\s+of\\s+(?:the\\s+)?(.+)", IntentAction::GET, std::nullopt},
            {"is\\s+(?:the\\s+)?(.+)\\s+(?:on|off)\\??", IntentAction::GET, std::nullopt},
        };
    }

    std::optional<std::string> resolveDevice(const std::optional<std::string>& device_name) {
        if (!device_name) {
            return std::nullopt;
        }

        std::string device_name_lower = toLower(trim(*device_name));

        // Exact match first
        auto it = device_aliases.find(device_name_lower);
        if (it != device_aliases.end()) {
            return it->second;
        }

        // Partial match
        for (const auto& [alias, device_id] : device_aliases) {
            if (device_name_lower.find(alias) != std::string::npos ||
                alias.find(device_name_lower) != std::string::npos) {
                return device_id;
            }
        }

        return std::nullopt;
    }

public:
    IntentParser(const std::string& config_path = "") {
        fs::path path;
        if (config_path.empty()) {
            path = fs::path(__FILE__).parent_path() / "config.json";
        } else {
            path = fs::path(config_path);
        }

        loadConfig(path);
    }

    Intent parse(const std::string& speech) {
        if (speech.empty()) {
            return Intent(IntentAction::UNKNOWN, std::nullopt, std::nullopt, speech);
        }

        std::string trimmed = trim(speech);

        // Try to match against action patterns
        for (const auto& pattern_obj : action_patterns) {
            try {
                std::regex pattern(pattern_obj.pattern, std::regex::icase);
                std::smatch match;

                if (std::regex_search(trimmed, match, pattern)) {
                    std::optional<std::string> device_name = std::nullopt;
                    if (match.size() > 1) {
                        device_name = trim(match[1].str());
                    }

                    std::optional<std::string> device_id = resolveDevice(device_name);
                    float confidence = device_id ? 0.9f : 0.6f;

                    return Intent(
                        pattern_obj.action,
                        device_id,
                        pattern_obj.command,
                        trimmed,
                        confidence
                    );
                }
            } catch (const std::regex_error& e) {
                std::cerr << "Regex error: " << e.what() << "\n";
                continue;
            }
        }

        // No pattern matched
        return Intent(IntentAction::UNKNOWN, std::nullopt, std::nullopt, trimmed, 0.0f);
    }

    void addDeviceAlias(const std::string& spoken_name, const std::string& device_id) {
        device_aliases[toLower(spoken_name)] = device_id;
    }
};

// Global parser instance
static std::unique_ptr<IntentParser> g_parser = nullptr;

IntentParser* getParser() {
    if (!g_parser) {
        g_parser = std::make_unique<IntentParser>();
    }
    return g_parser.get();
}

Intent parseIntent(const std::string& speech) {
    return getParser()->parse(speech);
}

// Main test function
int main() {
    IntentParser parser;

    std::vector<std::string> test_cases = {
        "Turn on the bedroom lights",
        "Switch off the living room lamp",
        "Toggle the kitchen light",
        "Turn on plug one",
        "What's the status of the bedroom light?",
        "Is the kitchen light on?",
        "Turn on the bathroom fan",  // Unknown device
        "Play some music",  // Unknown action
    };

    std::cout << "Intent Parser Test Cases:\n\n";
    for (const auto& speech : test_cases) {
        Intent intent = parser.parse(speech);
        std::cout << "Speech: \"" << speech << "\"\n";
        std::cout << "  → " << intent.toString() << "\n\n";
    }

    return 0;
}
